pkg/schedule/affinity/group.go

```diff
@@ -0,0 +1,381 @@
+// Copyright 2025 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package affinity
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"slices"
+	"time"
+
+	"github.com/tikv/pd/pkg/core"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/pkg/schedule/labeler"
+	"github.com/tikv/pd/pkg/slice"
+	"github.com/tikv/pd/pkg/utils/keyutil"
+)
+
+// condition is an enumeration that includes both Store state and Group state.
+type condition int
+
+const (
+	// groupAvailable indicates that the current Group allows affinity scheduling and disallows other balancing scheduling.
+	groupAvailable condition = iota
+
+	// groupDegraded indicates that the current Group does not generate affinity scheduling but still disallows other balancing scheduling.
+	// All values greater than groupAvailable and less than or equal to groupDegraded represent groupDegraded states.
+	// The groupDegraded state should have an expiration time. After it expires, it should be treated as groupExpired.
+	storeEvictLeader
+	storeDisconnected
+	storePreparing
+	storeLowSpace
+	groupDegraded
+
+	// groupExpired indicates that the current Group does not generate affinity scheduling and allows other balancing scheduling.
+	// All values greater than groupDegraded and less than or equal to groupExpired represent groupExpired states.
+	storeDown
+	storeRemovingOrRemoved
+	groupExpired
+
+	// groupAvailable, groupDegraded, and groupExpired define the Group’s availability lifecycle.
+	// Roughly:
+	//   groupAvailable ──degraded (e.g. store evict-leader)───────────────> groupDegraded
+	//   groupDegraded  ──recovered────────────────────────────────────────> groupAvailable // expected to be temporary
+	//   groupDegraded  ──expired──────────────────────────────────────────> groupExpired
+	//   groupAvailable ──directly failed (e.g. store removed)─────────────> groupExpired
+	//   groupExpired   ──reconfigured (e.g. peers moved to healthy stores)→ groupAvailable
+	// groupDegraded is intended to be a temporary state that may return to groupAvailable,
+	// while groupExpired usually represents a terminal state under the current topology,
+	// but can become groupAvailable again after the Group’s stores/peers are reconfigured.
+	// groupDegraded has an expiration time (degradedExpiredAt); once it expires, the Group is
+	// automatically treated as groupExpired.
+)
```

Comment:
Why not use a structure(contain storeState and groupState) instead of one type?

Reply:
Good